[MIG] delivery_mrw: Migration to 13.0

cubells opened the pull request.

pedrobaeza commented:
El nombre del módulo en el título no es correcto, no?

mariadforgeflow committed:
[ADD] delivery_mrw

victoralmau commented on delivery_mrw/models/delivery_carrier.py:
```diff
@@ -0,0 +1,259 @@
+    mrw_margin_left = fields.Integer(string="Margin Left", default=650)
+
+    def _get_mrw_wsdl_file(self):
+        wsdl_file = "mrw-api-test.wsdl"
+        if self.prod_environment:
+            wsdl_file = "mrw-api-prod.wsdl"
+        return wsdl_file
```
```suggestion
        return "mrw-api-prod.wsdl" if self.prod_environment else "mrw-api-test.wsdl"
```

victoralmau commented on delivery_mrw/models/delivery_carrier.py:
```diff
@@ -0,0 +1,259 @@
+        return wsdl_file
+
+    def _get_mrw_wsdl_tracking_file(self):
+        wsdl_file = "mrw-api-tracking-prod.wsdl"
+        return wsdl_file
```
```suggestion
        return "mrw-api-tracking-prod.wsdl"
```

victoralmau commented on delivery_mrw/models/delivery_carrier.py:
```diff
@@ -0,0 +1,259 @@
+        add them to the message.
+        """
+        wsdl_file = self._get_mrw_wsdl_file()
+        mrw_request = MrwRequest(
+            wsdl_file,
+            self.mrw_api_franchise,
+            self.mrw_api_subscriber,
+            self.mrw_api_user,
+            self.mrw_api_password,
+            self.mrw_api_dept_code,
+        )
```
```suggestion
        mrw_request = MrwRequest(self)
```

victoralmau commented on delivery_mrw/models/mrw_request.py:
```diff
@@ -0,0 +1,500 @@
+       the provided API. We leave the operations empty for future.
+    """
+
+    def __init__(self, wsdl_file, franchise, subscriber, user_id, password, dept_code):
+        """As the wsdl isn't public, we have to load it from local"""
+        self.wsdl_path = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "../api/%s" % wsdl_file
+        )
+        self.franchise = franchise or ""
+        self.subscriber = subscriber
+        self.mrw_user_id = user_id
+        self.mrw_pass = password
+        self.dept_code = dept_code or ""
```
```suggestion
    def __init__(self, carrier):
        """As the wsdl isn't public, we have to load it from local"""
        wsdl_file = carrier._get_mrw_wsdl_file()
        self.wsdl_path = os.path.join(
            os.path.dirname(os.path.realpath(__file__)), "../api/%s" % wsdl_file
        )
        self.franchise = carrier.mrw_api_franchise or ""
        self.subscriber = carrier.mrw_api_subscriber
        self.mrw_user_id = carrier.mrw_api_user
        self.mrw_pass = carrier.mrw_api_password
        self.dept_code = carrier.mrw_api_dept_code or ""
```

victoralmau commented on delivery_mrw/models/delivery_carrier.py:
```diff
@@ -0,0 +1,259 @@
+        }
+
+    def mrw_request_label(self, tracking_number):
+        wsdl_file = self._get_mrw_wsdl_file()
+        mrw_request = MrwRequest(
+            wsdl_file,
+            self.mrw_api_franchise,
+            self.mrw_api_subscriber,
+            self.mrw_api_user,
+            self.mrw_api_password,
+            self.mrw_api_dept_code,
+        )
```
```suggestion
        mrw_request = MrwRequest(self)
```

victoralmau commented on delivery_mrw/models/delivery_carrier.py:
```diff
@@ -0,0 +1,259 @@
+        self.ensure_one()
+        if not picking.carrier_tracking_ref:
+            return
+        wsdl_file = self._get_mrw_wsdl_tracking_file()
+        mrw_request = MrwRequest(
+            wsdl_file,
+            self.mrw_api_franchise,
+            self.mrw_api_subscriber,
+            self.mrw_api_user,
+            self.mrw_api_password,
+            self.mrw_api_dept_code,
+        )
```
```suggestion
        mrw_request = MrwRequest(self)
```

victoralmau commented on delivery_mrw/models/mrw_request.py:
```diff
@@ -0,0 +1,500 @@
+        """Get just tracking states from MRW info for the given reference"""
+        try:
+            data = self._prepare_getenvios(**vals)
+            _logger.debug(data)
+            transport = ZeepTransport(timeout=10)
+            zeep_client = ZeepClient(self.wsdl_path, transport=transport)
+            response = zeep_client.service.GetEnvios(**data)
+            _logger.debug(response)
```
```suggestion
            transport = ZeepTransport(timeout=10)
            zeep_client = ZeepClient(self.wsdl_path, transport=transport)
            response = zeep_client.service.GetEnvios(**data)
```

victoralmau commented on delivery_mrw/models/stock_picking.py:
```diff
@@ -0,0 +1,31 @@
+        for picking in self:
+            picking.has_mrw_carrier = (
+                picking.carrier_id and picking.carrier_id.delivery_type == "mrw"
+            )
```
```suggestion
```

victoralmau commented on delivery_mrw/views/stock_picking_views.xml:
```diff
@@ -0,0 +1,22 @@
+        <field name="arch" type="xml">
+            <xpath expr="//div[@name='tracking']" position="after">
+                <field name="has_mrw_carrier" invisible="1" />
+            </xpath>
```
```suggestion
```

victoralmau commented on delivery_mrw/views/stock_picking_views.xml:
```diff
@@ -0,0 +1,22 @@
+                    name="mrw_get_label"
+                    string="MRW Label"
+                    type="object"
+                    attrs="{'invisible':['|', ('has_mrw_carrier', '=', False), ('state', '!=', 'done')]}"
```
```suggestion
                    attrs="{'invisible':['|', ('carrier_id.delivery_type', '!=', 'mrw'), ('state', '!=', 'done')]}"
```

victoralmau commented on delivery_mrw/models/mrw_request.py:
```diff
@@ -0,0 +1,500 @@
+            self.client.set_options(soapheaders=(mrw_authinfo))
+            res = self.client.service.TransmEnvio(xml)
+        except Exception as e:
+            raise UserError(
```
¿Podrías implementar algo similar a esto https://github.com/OCA/l10n-spain/blob/13.0/delivery_seur/models/seur_request.py#L79-L81 en los errores de las peticiones para que se guarden los registros si se activa la opción debug para el transportista en cuestión?

mariadforgeflow commented:
Hola @cubells! 

Nosotros en ForgeFlow desarrollamos hace un tiempo el módulo de MRW para un cliente que ya esta usándolo en producción y no nos dimos cuenta que se abrió esta PR. He revisado el módulo y respecto al que desarrollamos nosotros (https://github.com/OCA/l10n-spain/pull/2482) faltarían los aspectos que detallo a continuación:

- Implementar la operación CancelarEnvio
- Obtener un Manifiesto con todos los envíos generados a través de Odoo, la información que se imprime es la misma que se obtiene al imprimirlo desde la Área de Clientes de MRW.
- No se informan los campos referentes a la dirección de envío: 
  -  CodigoTipoVia
  - Numero
  - Resto (donde según la documentación de desarrollador de MRW va el número piso, puerta etc)

- Se tendrían que restringir la creación de envíos  que no provengan de España y no sean direcciones de entrega nacionales. Según la documentación el método TransmEnvio solo permite envíos nacionales (se consideran nacionales a España, Portugal y Gibraltar). 

- Otros ampos que faltaría rellenar en la request de TransmEnvio:
  
  - Importe reembolso
  - Observaciones (nosotros cogemos el campo note del picking)

- Nosotros quitamos la opción de seleccionar invoice_policy ya que no aplica para este método de envío. Asimismo, ahora en el pedido el precio del delivery_product se coge 0 y al hacer update del precio salta NotImplementedError. Nosotros cogemos el precio del delivery_product directamente.

- He visto que si no hay un número de teléfono en el contacto del destinatario salta un error. "AttributeError: 'bool' object has no attribute 'replace'"
- En el picking al clicar en el botón de tracking salta: 
![image](https://user-images.githubusercontent.com/85128566/191724649-fafe6524-5408-41dd-9151-02f1fcb8e151.png)
nosotros redirigimos a la web de MRW con el seguimiento.

En el módulo que hemos desarrollado nosotros solo nos faltaría añadir la operación GetEnvios para tener el estado del envío en odoo.

Nosotros proponemos dos opciones:
- Nosotros añadimos a nuestro módulo la operación GetEnvios para el tracking y hacemos backport.
- Añades tú los cambios que hemos listado y haces un forwardport. 

Ya nos dices cómo prefieres avanzar.

Gracias!!

cubells commented:
Buah! perfecto @mariadforgeflow 

Me pongo a hacer los cambios y probarlos

mariadforgeflow commented:
Hola @cubells! Has podido mirarlo? Nos urge un poco tener el módulo mergeado. 
Si no, siempre está la opción que nosotros hagamos el backport de lo que tenemos en v14, ya que solo le faltaría añadir la operación GetEnvios.

cubells commented:
Hola @mariadforgeflow 

Tengo una posible ventana libre de tiempo para acabarlo este fin de semana pero no es seguro.

Si tanto os urge adelante con las modificaciones. Lo siento pero ahora tengo un pico grande de faena.

mariadforgeflow commented:
Hola @cubells, finalmente hemos añadido las modificaciones nosotros en https://github.com/OCA/l10n-spain/pull/2566, si pudieras revisarlo nos sería de ayuda.
Gracias!

cubells commented:
@mariadforgeflow claro.

Me lo apunto para el fin de semana, haré yo un backport de esas modificaciones para que la versión 12 tenga las mismas opciones

mariadforgeflow commented:
@cubells perfecto gracias! De momento solo está la PR en la v14 por si se tuviera que cambiar algo, y ya después haríamos el backport a la 13.0:)

Commit:
[UPD] Update delivery_mrw.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

Commit:
[IMP] delivery_mrw: Add 0205 (Urgente 19 Expediciones) service

Commit:
[UPD] Update delivery_mrw.pot

OCA-git-bot committed:
delivery_mrw 14.0.1.1.0

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-14.0/l10n-spain-14.0-delivery_mrw
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-14-0/l10n-spain-14-0-delivery_mrw/

cubells committed:
[IMP] delivery_mrw: black, isort, prettier

cubells committed:
[MIG] delivery_mrw: Migration to 13.0
